compiler/rustc_interface/src/passes.rs

```diff
@@ -336,6 +337,20 @@ fn early_lint_checks(tcx: TyCtxt<'_>, (): ()) {
     )
 }
 
+fn env_var(tcx: TyCtxt<'_>, key: Symbol) -> Option<Symbol> {
+    let var = match std::env::var(key.as_str()) {
+        Ok(var) => Some(Symbol::intern(&var)),
+        Err(VarError::NotPresent) => None,
+        Err(VarError::NotUnicode(var)) => {
+            tcx.dcx().emit_err(errors::EnvVarNotUnicode { key, var });
+            None
+        }
+    };
+    // Also add the variable to Cargo's dependency tracking
+    tcx.sess.psess.env_depinfo.borrow_mut().insert((key, var));
```

Comment:
Hmm, a complication here is that this query will only work with the dependency info files that Cargo reads if it's called _before_ `passes::write_dep_info` at: https://github.com/rust-lang/rust/blob/a3f76a26e045a760bb1163b7eab36872985242d5/compiler/rustc_driver_impl/src/lib.rs#L437

So everything done during `passes::analysis` and `Linker::codegen_and_build_linker` will not be able to use this to, and will need some other mechanism for populating `ParseSess::env_depinfo`.

Is it possible for us to move the `write_dep_info` pass to the very end, or just before linking? Or is it important that it happens as soon as possible for parallelizing Cargo, or something?

Reply:
>Or is it important that it happens as soon as possible for parallelizing Cargo, or something?

It is important for depinfo to be generated as early as possible, without performing full compilation.
All *file* dependencies in particular, and env var dependencies from `env!` and proc macros are ready after name resolution is completed.

I think it's fine to still lose some env depinfo in this PR, it will still better than what we have now.
Could you add a comment to this code, explaining that env vars added after the call to `write_dep_info` will be lost from the dep file?

(We can think what to do with the Apple env vars specifically later.)

Reply:
I have added a comment to the `env_var` query, and added `rustc_codegen_ssa::register_environment_variables` as an extension point for where to put this. Suggestions very welcome on how to do it differently!